course3/homework-string/my_string.cpp

```diff
@@ -0,0 +1,209 @@
+#include "my_string.h"
+#include <memory>
+
+        MyString::MyString()
+        {
+            mBuffer=nullptr;
+        }
+
+        MyString::MyString(const char* other)
+        {
+            mBuffer = std::make_unique<char[]>(std::strlen(other) + 1);
+            std::strcpy(mBuffer.get(), other);
+        }
+
+        MyString::MyString(const MyString& other)
+        {
+            mBuffer = std::make_unique<char[]>(other.size() + 1);
+            std::strcpy(mBuffer.get(), other.mBuffer.get());
+        }
+
+        std::size_t MyString::size() const
+        {
+            if(!mBuffer)
+                return 0;
+
+            return std::strlen(mBuffer.get());
+        }
+
+        bool MyString::empty() const
+        {
+            return mBuffer == nullptr || mBuffer.get()[0] == 0;
+        }
+
+        const char* MyString::c_str() const
+        {
+            if(!mBuffer)
+                return "";
+
+            return mBuffer.get();
+        }
+
+        char MyString::operator[](std::size_t index) const
+        {
+            return mBuffer.get()[index];
+        }
+        char& MyString::operator[](std::size_t index)
+        {
+            return mBuffer.get()[index];
+        }
+
+        bool MyString::operator==(const MyString& other) const
+        {
+            if(this -> empty() && other.empty())
+                //TODO: this is not completely right just yet. we need nullptr checks for later on.
+                return true;
+
+            return std::strcmp(mBuffer.get(), other.mBuffer.get()) == 0;
+        }
+        bool MyString::operator!=(const MyString& other) const
+        {
+            return !(*this==other);
+        }
+
+        bool MyString::operator<(const MyString& other) const
+        {
+            if(this->empty() && other.empty())
+                return false;
+
+            return std::strcmp(this->c_str(), other.c_str()) < 0;
+        }
+
+        bool MyString::operator<=(const MyString& other) const
+        {
+            if(this->empty() && other.empty())
+                return true;
+
+            return std::strcmp(this->c_str(), other.c_str()) <= 0;
+        }
+
+        bool MyString::operator>(const MyString& other) const
+        {
+            if(this->empty() && other.empty())
```

Comment:
There's no point implementing all the other operators, ```<``` and ```==``` are all you ever need, the rest you _could_ implement in terms of the two, but you wouldn't have to, because you could use Boost.Operators to do that for you.

http://www.boost.org/doc/libs/1_59_0/libs/utility/operators.htm